libs/partners/google-vertexai/langchain_google_vertexai/chat_models.py

```diff
@@ -444,19 +473,36 @@ async def _agenerate(
             # set param to `functions` until core tool/function calling implemented
             raw_tools = params.pop("functions") if "functions" in params else None
             tools = _format_tools_to_vertex_tool(raw_tools) if raw_tools else None
-            response = await chat.send_message_async(
-                message,
-                generation_config=params,
-                tools=tools,
-                safety_settings=safety_settings,
-            )
-            generations = [
-                ChatGeneration(
-                    message=_parse_response_candidate(c),
-                    generation_info=get_generation_info(c, self._is_gemini_model),
+            try:
+                response = chat.send_message_async(
+                    message,
+                    generation_config=params,
+                    tools=tools,
+                    safety_settings=safety_settings,
                 )
-                for c in response.candidates
-            ]
+            except ResponseBlockedError as e:
+                response = e.responses[0]
+            generations = []
+            for c in response.candidates:
```

Comment:
it looks like quite a lot of code duplication. can it be moved to a separate function maybe?